url7f43a4752f_shalynnho_pokemonfactory_tgz-pJ8-Networking_ServerJ8/src/agent/Agent.java

```diff
@@ -0,0 +1,155 @@
+	}
+
+	/** Print message */
+	protected void print(String msg) {
+		print(msg, null);
+	}
+
+	/** Print message with exception stack trace */
+	protected void print(String msg, Throwable e) {
+		StringBuffer sb = new StringBuffer();
+		sb.append("[Agent]");
+		sb.append(getName());
+		sb.append(": ");
+		sb.append(msg);
+		sb.append("\n");
+		if (e != null) {
+			sb.append(StringUtil.stackTraceString(e));
+		}
+
+		// if (this.getClass() != agent.FCSAgent.class // && this.getClass() != agent.StandAgent.class
+		// && this.getClass() != agent.KitRobotAgent.class && this.getClass() != agent.ConveyorAgent.class
+		// && this.getClass() != agent.LaneAgent.class && this.getClass() != agent.GantryAgent.class
+		// && this.getClass() != agent.CameraAgent.class */&& this.getClass() != agent.NestAgent.class
+		// && this.getClass() != agent.FeederAgent.class // && this.getClass() != agent.PartsRobotAgent.class
+		// && this.getClass() != agent.test.mock.MockGraphics.class) {
+		System.out.print(sb.toString());
+		// }
+
+		if (console != null) {
+			console.sendMessage(getName(), msg);
+		}
+	}
+
+	/**
+	 * Sets the graphical representation of this agent.
+	 * 
+	 * @param dg
+	 */
+	public abstract void setGraphicalRepresentation(DeviceGraphics dg);
+
+	public void setConsoleWriter(ConsoleWriter cw) {
+		console = cw;
+	}
+
+	/** Start agent scheduler thread. Should be called once at init time. */
+	public synchronized void startThread() {
+		print("Thread started.");
+		if (agentThread == null) {
+			agentThread = new AgentThread(getName());
+			agentThread.start(); // causes the run method to execute in the
+								 // AgentThread below
+		} else {
+			agentThread.interrupt();// don't worry about this for now
+		}
+	}
+
+	/** Stop agent scheduler thread. */
+	// In this implementation, nothing calls stopThread().
+	// When we have a user interface to agents, this can be called.
+	public void stopThread() {
+		print("Thread stopped");
+		if (agentThread != null) {
+			agentThread.stopAgent();
+			agentThread = null;
+		}
+	}
+
+	/**
+	 * Agent scheduler thread, calls respondToStateChange() whenever a state change has been signalled.
+	 */
+	private class AgentThread extends Thread {
+		private volatile boolean goOn = false;
+
+		private AgentThread(String name) {
+			super(name);
+		}
+
+		@Override
+		public void run() {
+			goOn = true;
+
+			while (goOn) {
+				try {
+					// The agent sleeps here until someone calls,
+					// stateChanged(),
+					// which causes a call to stateChange.give(), which wakes up
+					// agent.
+					stateChange.acquire();
+					// The next while clause is the key to the control flow.
+					// When the agent wakes up it will call
+					// respondToStateChange()
+					// repeatedly until it returns FALSE.
+					// You will see that pickAndExecuteAnAction() is the agent
+					// scheduler.
+					// print("In scheduler.");
+					while (pickAndExecuteAnAction()) {
+						;
+					}
+				} catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)